docs/proposals/node-allocatable.md

```diff
@@ -0,0 +1,177 @@
+   "/kube-proxy", "/system" etc. raw containers for monitoring system component resource usage
+   patterns and detecting regressions. Eventually we want to cap system component usage to a certain
+   limit / request. However this is not currently feasible because 1) docker still uses tons of
+   computing resources and 2) we have not yet defined the minimal system requirements, so we cannot
+   control Kubernetes nodes or know about arbitrary daemons, which can make the system resources
+   unmanageable. Even with a resource cap we cannot do a full resource management on the node, but
+   with the proposed parameters we can mitigate really bad resource overcommits.
+2. For external schedulers (such as mesos, hadoop, etc.) integration, they might want to partition
+   compute resources on a given node, limiting how much Kubelet can use. We should provide a
+   mechanism by which they can query kubelet, and reserve some resources for their own purpose.
+
+### Scope of proposal
+
+This proposal deals with resource reporting through the [`Allocatable` field](#allocatable) for more
+reliable scheduling, and minimizing resource over commitment. This proposal *does not* cover
+resource usage enforcement (e.g. limiting system component usage), and pod eviction (e.g. when
+reservation grows).
+
+## Design
+
+### Definitions
+
+![image](node-allocatable.png)
+
+1. **Machine Capacity** - Already provided as
+   [`NodeStatus.Capacity`](https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/v1/definitions.html#_v1_nodestatus),
+   this is total capacity read from the machine instance, and assumed to be constant.
+2. **Reserved** (proposed) - Compute resources reserved for non-Kubernetes and non-system
+   workloads. For example, processes scheduled to run alongside Kubernetes on Mesos.
+3. **Kubelet Allocatable** - Compute resources available for scheduling (including scheduled &
+   unscheduled resources). This value is currently implicit, and the focus of this proposal. See
+   [below](#api-changes) for more details.
+4. **System Reserved** (proposed) - Compute resources reserved for system components such as the docker daemon, kubelet, kube proxy, etc.
+
+### API changes
+
+#### Allocatable
+
+Add `Allocatable` (4) to
+[`NodeStatus`](https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/v1/definitions.html#_v1_nodestatus):
+
+```
+type NodeStatus struct {
+  ...
+  // Allocatable represents schedulable resources of a node.
+  Allocatable ResourceList `json:"allocatable,omitempty"`
+  ...
+}
+```
+
+Allocatable will be computed by the Kubelet and reported periodically to the API server. It is
+defined to be:
+
+```
+   [Allocatable] = [Machine Capacity] - ( [Reserved] + [System Reserved] + [Kernel usage (smoothed)] )
+```
+
+The scheduler will use `Allocatable` in place of `Capacity` when scheduling pods, and the Kubelet
```

Comment:
We may need to recommend a safe upgrade path where both the Kubelets and scheduler(s) are updated before Allocatable is configured to be less than Capacity. Otherwise, schedulers using Capacity could incur many rejected pods.